[UI] Implement timeline component

yaooluu opened the pull request:
This change implements a static timeline component using HTML/CSS, based on the design spec from Clarity team.
    
Closes #3199
    
Signed-off-by: Yao Lu <ylu31@ncsu.edu>

Same code updated at: https://stackblitz.com/edit/clarity-timeline-demo-v3. Ready for reviewers to play with the code.

**Question list:**
- [ ] In demo project I'm adding the `spinner` icon as a custom icon. How to achieve the same for the production? i.e. component consumer should not need to add this icon by themselves?

gnomeontherun commented:
@colinreedmiller Here is the final implementation to review https://deploy-preview-3594--vmware-clarity.netlify.com/dev click on Timeline in sidenav. Do you approve?

Jinnie commented:
@gnomeontherun won't we need a website documentation page? Or we'll have a separate PR/ticket for it?

gnomeontherun commented on src/dev/src/app/timeline/timeline.demo.module.ts:
```diff
@@ -0,0 +1,20 @@
+import { CommonModule } from '@angular/common';
+import { NgModule } from '@angular/core';
+
+import { ClarityModule } from '@clr/angular';
```
nit: this isn't a component that uses Clarity Angular, don't need this technically.

yaooluu replied:
Removed.

yaooluu replied:
Switched to ClrIconModule as I'm using `clr-icon` in the component.

gnomeontherun commented on src/dev/src/app/timeline/timeline.demo.ts:
```diff
@@ -0,0 +1,27 @@
+export class TimelineDemo {
+  ngOnInit() {
+    ClarityIcons.add({
+      spinner: `<svg xmlns="http://www.w3.org/2000/svg" width="36" height="36" viewBox="0 0 36 36"><title>spinner 23</title><rect width="36" height="36" fill-opacity="0"/><circle cx="18" cy="18" r="14.85" fill="none" stroke="#000" stroke-miterlimit="10" stroke-opacity="0.15" stroke-width="2.3"/><path d="M8.24,29.2A14.86,14.86,0,0,1,18,3.15" fill="none" stroke="#007cbb" stroke-miterlimit="10" stro ...
```
For this to work, it seems we need to have a spinner icon, and not have it hard coded here. Can we add this to our icon core library? @Shijir does that make sense?

yaooluu replied:
@Shijir Though in our previous conversation we discussed that the icon could be added using `ClarityIconApi` by our own, we still need it to be part of Clarity official icon so that consumers using static version of timeline can see the icon. Is that correct?

Shijir replied:
This icon was supposed to be added into the module of this component. Client apps that don't have access to this icon won't be able to render this icon.

gnomeontherun replied:
Ok we'll leave this for now, but need to figure this out before we announce this functionality.

gnomeontherun commented on src/dev/src/app/timeline/timeline.demo.routing.ts:
```diff
@@ -0,0 +1,23 @@
+    component: TimelineDemo,
+    children: [
+      { path: '', redirectTo: 'timeline-types', pathMatch: 'full' },
+      { path: 'timeline-types', component: TimelineTypesDemo },
```
This can be the blank path route and remove the redirect above. We don't need two routes.

yaooluu replied:
Great suggestion, removed all routes make it one component only. Followed another example before and thought there might be multiple demo pages required.

gnomeontherun commented on src/dev/src/app/timeline/timeline.demo.ts:
```diff
@@ -0,0 +1,27 @@
+
+@Component({
+  selector: 'clr-timeline-demo',
+  styleUrls: ['./timeline.demo.scss'],
```
The `selector` and `styleUrls` can be removed, as well as the timeline.demo.scss file since its empty.

yaooluu replied:
All removed.

gnomeontherun commented on src/dev/src/app/timeline/timeline.demo.ts:
```diff
@@ -0,0 +1,27 @@
+@Component({
+  selector: 'clr-timeline-demo',
+  styleUrls: ['./timeline.demo.scss'],
+  template: `
```
Let's remove this inline template, and use an HTML file with the `timeline-types.html` content. Then you can simplify to just have one component instead of two. See the `radios` demo directory for an example.

yaooluu replied:
Done.

gnomeontherun commented on src/clr-angular/timeline/_timeline.clarity.scss:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  // specific styles for horizontal timeline
+  .clr-timeline.clr-timeline-horizontal {
```
I think we should assume horizontal is default, so they shouldn't have to add `.clr-timeline-horizontal` to achieve that.

yaooluu replied:
After taking a look at some other component at Clarity, I agree it's the right solution to actually default to horizontal. I'll work on merge the styles.

gnomeontherun commented on src/clr-angular/timeline/_timeline.clarity.scss:
```diff
@@ -0,0 +1,162 @@
+// This software is released under MIT license.
+// The full license information can be found in LICENSE in the root directory of this project.
+
+@include exports('timeline.clarity') {
```
I don't see any dark theme styles here, has this been considered or spec'd out?

yaooluu replied:
Unfortunately it was never in the spec.

gnomeontherun commented on src/clr-angular/timeline/_timeline.clarity.scss:
```diff
@@ -0,0 +1,162 @@
+
+  .clr-timeline-step {
+    display: flex;
+    position: relative;
```
This might be a problem, it will cause clipping on anything like a dropdown menu. Have you tried any other options that would get around this? @hippee-lee any additional insights based on popover work?

yaooluu replied:
I did some experiments and the horizontal timeline still works fine without step (relative) and step-body:before (absolute).

But if I remove those, the vertical timeline doesn't work. The "line" is not displayed at all between steps. Or with only step-body::before (absolution) it won't be able to get its parent's (timeline-step) height by using 100%, instead it get's the full height of the entire timeline..Not sure what should be the solution to get rid of it...

colinreedmiller commented:
@gnomeontherun the original spec had no dark theme design. My work  has been to improve the visual design such that it was clearer - the decisions about what content to support were part of the original design. We need to look at use cases since from what I understand, this component is not widely used in VMWare products and its original need has been superseded by additional datagrid functionality.

yaooluu commented:
@gnomeontherun Could you review the above changes again and let me know your thoughts?

colinreedmiller commented:
I asked @lil-kim to take a look at the design spec and add the required callouts for dark theme so we have this available while its in flight.

Figma file here: https://www.figma.com/file/l7MoP6qowUUqMgNVWnxS4g/Timeline-component-Static-3199?node-id=12%3A2030

gnomeontherun commented:
I'm going to review again, wasn't sure if you were done. @colinreedmiller I think we can handle the dark theme version of this separately to help close the loop on this PR.

yaooluu commented:
> I'm going to review again, wasn't sure if you were done. 
I addressed most comments except for the `position: relative`, which I could find alternative that won't mess up the layout. Would appreciate some help here.

colinreedmiller commented:
@gnomeontherun - Yes. I am completely finished with Light theme and its approved - implementation is correct. I included the dark theme callouts in case it was preferable to build it at the same time but later is fine also. Closing this PR without dark theme is fine with me.

gnomeontherun commented:
@yaooluu Can you squash your commits and update gemini tests? Since you added something to the dev app some of the images that include the nav get flagged.

gnomeontherun commented:
Also make sure your commit use signed-off-by!

gnomeontherun commented:
Just following up on this @yaooluu do you have an ETA on getting the PR fixed up?

yaooluu commented:
> Just following up on this @yaooluu do you have an ETA on getting the PR fixed up?

Yes I want to get this done before end of week. I actually has some questions regarding squash, rebase, and the `gemini` tests as I'm not familiar with the process. I'll reach out to the team offline to get it resolved.

yaooluu commented:
@gnomeontherun I've squashed my commits and updated the gemini test. The pipeline is passing now!

gnomeontherun commented:
Thanks Yao! We'll get it reviewed and merged soon!
